File: Go/Day07.go
package main

import AH "./adventhelper"

import (
	"strconv"
	"strings"
)

type Node struct {
    name string
    size int
    file bool
    parent *Node
    children []*Node
}


func (n *Node) dirSize() int {
		if n.file {
			return n.size
		} else {
			subsum := 0
			for _, c := range n.children {
				subsum += c.dirSize()
			}
    	return subsum
		} 
}

func changeDir(s string, n *Node) (*Node) {
	if s == ".." {
		return n.parent
	}

	for _, c := range n.children {
		if c.name == s {

			return c
		}
	}

	return n
}

func parseCLI(ss []string, n *Node) {
	ptr := n
	for _, s := range ss {
		if s == "$ ls" { // list everything in directory
			// nothing ? 
		} else if (s[0:4] == "$ cd") { // change pointer
			data := strings.Split(s, " ")
			ptr = changeDir(data[2], ptr)
		} else { // file / dir
			data := strings.Split(s, " ")
			if data[0] == "dir" { // is directory
				n_new := Node {name:data[1], size:0, file:false, parent:ptr, children:[]*Node{}}
				ptr.children = append(ptr.children, &n_new)
			} else { // is file
				fsize, _:= strconv.Atoi(data[0])
				n_new := Node {name:data[1], size:fsize, file:true, parent:ptr, children:[]*Node{}}
				ptr.children = append(ptr.children, &n_new)				
			}
		}
	}
}

func (n *Node)part1(size int) (int) {
	counter := 0
	if n.file {
		return 0
	} else {
		if n.dirSize() < size {
			counter += n.dirSize()
		}
	}

	for _, c := range n.children {
		counter += c.part1(size)
	}

	return counter
}

func (n *Node)part2(m map[string]int) {
	if n.file {
		return
	} else {
		m[n.name] = n.dirSize()
	}

	for _, c := range n.children {
		c.part2(m)
	}

	return
}

func main() {
	cli_instructions, _ := AH.ReadStrFile("../input/input07.txt")
	root := Node {name:"root", size:0, file:false, parent:nil, children:[]*Node{}}

	parseCLI(cli_instructions[1:], &root)

	toDelete := root.dirSize() - 40000000

	dirSizes := make(map[string]int)
	root.part2(dirSizes)

	part2 := root.dirSize()
	for _, v := range dirSizes {
		if (v > toDelete) && (v < part2) {
			part2 = v
		}
	}

	AH.PrintSoln(7, root.part1(100000), part2)

	return
}

File: C++/Day07.cpp
#include "AH.h"

namespace Day07
{

	struct Node
	{
		std::string name;
		unsigned int size;
		bool is_file;
		Node * parent;
		std::vector<Node *> children;

		int dirSize() const;
		Node * changeDir(const std::string s);
		void addNode(const std::string name, const unsigned int size, const bool is_file);
		void getDirSizes(std::map<std::string, unsigned int> & m) const;
	};

	int Node::dirSize() const
	{
		if (is_file)
		{
			return size;
		}
		else
		{
			unsigned int sum = 0;
			for (const auto c : children)
				sum += c->dirSize();

			return sum;
		}
	}

	Node * Node::changeDir(const std::string s)
	{
		if (s == "..")
			return parent;

		for (const auto c : children)
			if (c->name == s)
				return c;

		return this;
	}

	void Node::addNode(const std::string name, const unsigned int size, const bool is_file)
	{
		Node * n = new Node;
		n->name = name;
		n->size = size;
		n->is_file = is_file;
		n->parent = this;
		n->children.clear();
		children.push_back(n);
	}

	void parseCLI(const std::vector<std::string> ss, Node * n)
	{
		for (const auto s: ss)
		{
			if (s == "$ ls")
			{
				// do nothing
			}
			else if (s.substr(0, 4) == "$ cd")
			{
				const auto data = AH::Split(s, ' ');
				n = n->changeDir(data[2]);
			}
			else
			{
				const auto data = AH::Split(s, ' ');
				if (data[0] == "dir")
					n->addNode(data[1], 0, false);
				else
					n->addNode(data[1], std::stoi(data[0]), true);
			}
		}
	}

	void Node::getDirSizes(std::map<std::string, unsigned int> & m) const
	{
		if (is_file)
			return;
		
		std::string full_name = "root";
		if (parent != nullptr)
			full_name = parent->name + "|" + name;

		m[full_name] = dirSize();

		for (const auto c: children) 
			c->getDirSizes(m);

		return;
	}

	int Run(const std::string& filename)
	{
		auto lines = AH::ReadTextFile(filename);
		lines.erase(lines.begin());
		Node root = Node();
		root.name = "root";
		root.size = 0;
		root.is_file = false;
		root.parent = nullptr;
		root.children.clear();

		std::map<std::string, unsigned int> m;

		parseCLI(lines, & root);
		root.getDirSizes(m);

		unsigned int part1 = 0;
		unsigned int part2 = m["root"];
		const unsigned int toDelete = m["root"] - 40000000;

		for (auto const& [k, v] : m)
		{
			if ((v > toDelete) && (v < part2))
				part2 = v;

			if (v <= 100000)
				part1 += v;
		}

		AH::PrintSoln(7, part1, part2);

		return 0;
	}

}
